core/src/mast/multi_forest_node_iterator.rs

```diff
@@ -0,0 +1,508 @@
+        forest_b.add_node(block_foo).unwrap();
+        forest_b.make_root(id_call_b);
+
+        let nodes = MultiMastForestNodeIter::new(vec![&forest_a, &forest_b]).collect::<Vec<_>>();
+
+        assert_eq!(nodes.len(), 5);
+
+        // The replacement for the external node from forest A.
+        assert_eq!(
+            nodes[0],
+            MultiMastForestIteratorItem::Node {
+                forest_idx: 1,
+                node_id: MastNodeId::new_unchecked(2)
+            }
+        );
+        // The external node replaced by the block foo from forest B.
+        assert_eq!(
+            nodes[1],
+            MultiMastForestIteratorItem::ExternalNodeReplacement {
+                replacement_forest_idx: 1,
+                replacement_mast_node_id: MastNodeId::new_unchecked(2),
+                replaced_forest_idx: 0,
+                replaced_mast_node_id: MastNodeId::new_unchecked(0)
+            }
+        );
+        // The call from forest A.
+        assert_eq!(
+            nodes[2],
+            MultiMastForestIteratorItem::Node {
+                forest_idx: 0,
+                node_id: MastNodeId::new_unchecked(1)
+            }
+        );
+        // The replacement for the external node that is replaced by the Call in forest A.
+        assert_eq!(
+            nodes[3],
+            MultiMastForestIteratorItem::ExternalNodeReplacement {
+                replacement_forest_idx: 0,
+                replacement_mast_node_id: MastNodeId::new_unchecked(1),
+                replaced_forest_idx: 1,
+                replaced_mast_node_id: MastNodeId::new_unchecked(0)
+            }
+        );
+        // The call from forest B.
+        assert_eq!(
+            nodes[4],
+            MultiMastForestIteratorItem::Node {
+                forest_idx: 1,
+                node_id: MastNodeId::new_unchecked(1)
+            }
+        );
+    }
+
+    /// Tests that a node which appears twice in a tree is returned in the required order.
+    ///
+    /// In this test we have a MastForest with a tree like this:
+    ///
+    ///     3 <- Split Node
+    ///    / \
+    ///   1   2
+    ///  /     \
+    /// 0       0
+    ///
+    /// In a previous implementation we marked the nodes that we discovered as visited immediately
+    /// and did not add them again when encountering them again. This was not correct because of
+    /// this example. In the iterator, we need to discover nodes in reverse so we can later pop
+    /// them off the stack in postorder. So we first descend down child 2 of the split node.
+    /// This means we would mark 0 as visited and then add 2 to the stack. Then we descend down
+    /// child 1 but do not add 0 again because it is already visited and we add 1. So we end up
+    /// with this stack for the tree: [3, 2, 0, 1]. This means when we pop off nodes from the
+    /// stack we get to 1 before we get to 0 and that violates the guarantees from this
+    /// iterator.
```

Comment:
Question: if I understood correctly, in the current implementation we do put duplicates onto the stack but then when we pop them, we check if a node has been visited and skip of so - correct?

Reply:
Yeah, this was a late change because of this test: https://github.com/0xPolygonMiden/miden-vm/blob/310f90341c3b9128d3527c9d300d66dbf620f660/core/src/mast/multi_forest_node_iterator.rs#L463-L488

This is also related to @bitwalker's comment here: https://github.com/0xPolygonMiden/miden-vm/pull/1534#discussion_r1815422090
Meaning this change made the implementation a less efficient.
I'm looking into this right now. I don't understand the whole picture yet and if we need to use a topological ordering instead.

Reply:
Ah - I see. So, with the current approach we'd be inserting the same procedure more than once into the stack and that may in fact result in a lot of extra nodes - right?

Could we use something like `VecDeque` to avoid this problem? Basically, we traverse the tree in the "normal" order (i.e., visiting nodes in the following order `left 0`, `1`, `right 0`, `2`, `3`) and append the previously unvisited nodes to the end of the queue. But then we pop them from the front to return them from the iterator.

This is just an idea - I haven't really thought it through - so, not sure if it'll actually work.

Reply:
> Could we use something like `VecDeque` to avoid this problem?

Yes! I just tried exactly that locally and all existing tests pass but each node is only visited once.
The `VecDeque` also has the advantage of making it easier to understand - no reversing of operations needed in code or in documentation :smile:

Reply:
I guess one potential issue could be if we need to push onto the stack which is not empty - but I'm not sure if that happens in practice.

Reply:
> I guess one potential issue could be if we need to push onto the stack which is not empty - but I'm not sure if that happens in practice.

I'm not quite sure I understand what you mean.
A tree discovery would push elements onto the back of the deque and build up the entire tree that way.
Only then do we start popping them off the front until the deque is empty (one pop operation for each `next()` call).
Once the deque is empty, we start another discovery, followed by another series of pop operations, and so on.

Reply:
Right, in this scenario everything should work fine - i.e.:

- Start with empty queue
- Add nodes to the queue during tree discovery
- Drain the queue entirely
- Add nodes to the queue during tree discovery
- Drain the queue entirely
- etc.

But if we needed to add something to the queue before it was drained entirely, there could be an issue. Since it doesn't happen in practice, we should be OK.